apps/dashr-ocean-optics/app.R

```diff
@@ -0,0 +1,580 @@
+            improve if this feature is turned off, and that it will not be
+            possible to zoom in on any portion of the plot if it is turned
+            on.
+            '
+            )
+          )
+        )
+)))
+  
+app$layout(htmlDiv(id = 'main', children = pageLayout))
+
+
+############################
+# Helper Variables & FUNs
+############################
+
+# input list control elements
+controlValues <- lapply(1:length(controls), function(i) {
+  input(id = controls[[i]]$props$children[[2]]$props$id,
+        property = ifelse(
+          length(controls[[i]]$props$children[[2]]$props$value) > 0,
+          "value",
+          "on"
+        ))
+})
+
+# state list control elements
+controlValuesState <- lapply(1:length(controls), function(i) {
+  state(id = controls[[i]]$props$children[[2]]$props$id,
+        property = ifelse(
+          length(controls[[i]]$props$children[[2]]$props$value) > 0,
+          "value",
+          "on"
+        ))
+})
+
+# generates intensities
+SampleSpectrum <- function(scale, knobIntensity, x) {
+  scale * (exp(-1 * ( (x - 500) / 5) ** 2) +
+             0.01 * runif(length(x))) + knobIntensity * 10
+}
+
+
+############################
+# Callbacks
+############################
+
+# disable/enable the update button depending on whether options have changed
+app$callback(
+  output = list(id = "submit-button", property = "style"),
+  params = c(controlValues,
+             list(input(id = "submit-button", "n_clicks_timestamp"))),
+
+  function(...) {
+    args <- list(...)
+
+    argsTime <- args[[length(args)]]
+    now <- as.numeric(Sys.time()) * 1000
+
+    disabled <- list(
+      'color' = colors[['accent']],
+      'backgroundColor' = colors[['background']],
+      'cursor' = 'not-allowed'
+    )
+
+    enabled <- list(
+      'color' = colors[['background']],
+      'backgroundColor' = colors[['accent']],
+      'cursor' = 'pointer'
+    )
+
+    # if the button was recently clicked (less than a second ago), then
+    # it's safe to say that the callback was triggered by the button; so
+    # we have to "disable" it
+    if (now - argsTime < 500 && argsTime > 0) {
+      return(disabled)
+    } else {
+      return(enabled)
+    }
+  }
+)
+
+
+# disable/enable controls
+app$callback(
+  output = list(id = "controls", property = "children"),
+  params = list(input(id = "power-button", property = "on")),
+
+  function(pwr_on) {
+    lapply(1:length(controls),
+           function(i) {
+             controls[[i]]$props$children[[2]]$props$disabled <- !pwr_on
+             controls[[i]]
+           })
+  }
+)
+
+
+# disable/enable intensity knob
+app$callback(
+  output = list(id = "light-intensity-knob", property = "disabled"),
+  params = list(
+    input(id = "power-button", property = "on"),
+    input(id = "light-source-input", property = "value")
+  ),
+
+  function(pwr, ls) {
+    return(!(pwr && ls != "NULL" && ls != ""))
```

Comment:
Changed to:
```
  function(pwr, lsi) {
    return(!(pwr && !is.null(lsi[[1]]) && lsi != ""))
  }
)
```
Since `lsi` is type `list` when the `dccDropdown` is empty and `character` otherwise, had to change `lsi` to `lsi[[1]]`.